### ISSUE TITLE
[change] Only report first instance of unreachable code

### ISSUE THREAD

**Issue by adambuczynski** [eslint/eslint#6583]
In ESLint 2.11.1 (bundled with Atom linter-eslint), the error for `no-unreachable` marks every single line of code after the `return` statement. Maybe this is redundant and it should only mark the first instance of unreachable code? It's pretty obvious that everything after that line will also be unreachable.

Illustration:

![image](https://cloud.githubusercontent.com/assets/490562/16543831/b493f06c-4140-11e6-9b78-28f59571a005.png)

Perhaps we could control this via a flag/setting, in case some people do prefer to have every line marked.

**Comment by platinumazure** [comment]
I think it's worth discussing for sure. I think having it be an option will make the enhancement more palatable.

@eslint/eslint-team Thoughts?

**Comment by adambuczynski** [comment]
I find that one problem with marking every line was that it lead to "error blindness", e.g. I could not spot an actual error anymore because I was trying to ignore all the irrelevant duplicate errors.

**Comment by nzakas** [comment]
I can support this as an option, but think we should keep the current default behavior. :+1: 

@adambuczynski are you willing to implement this change if others agree?

**Comment by adambuczynski** [comment]
@nzakas yeah I could have a look

**Comment by mysticatea** [comment]
Hm...

At first, I'm sorry that I have not implemented #3307.
But what about that it merges warnings of continuous nodes?
If #3307 is implemented, a warning can show range. Then,

```js
var startNode = null;
var endNode = null;

function reportIfUnreachable(node) {
    if (node && currentCodePath.currentSegments.every(isUnreachable)) {
        // Store this node to check consecutive range.
        if (!startNode) {
            endNode = startNode = node;
            return;
        }

        // Skip if this node is inside of the current range.
        if (startNode.range[0] <= node.range[0] && node.range[1] <= endNode.range[1]) {
            return;
        }

        // Skip if the previous token of this node is inside of the current range.
        // This involves that this node and the current range are consecutive.
        var prev = context.getTokenBefore(node);
        if (startNode.range[0] <= prev.range[0] && prev.range[1] <= endNode.range[1]) {
            endNode = node;
            return;
        }
    }

    // Report the current range since this node and the current range are not consecutive.
    if (startNode) {
        context.report({
            message: "Unreachable code.",
            loc: {start: startNode.loc.start, end: endNode.loc.end},
            node: startNode
        });
    }

    // Update the current range.
    startNode = endNode = node;
}
```

**Comment by nzakas** [comment]
@mysticatea I think the problem with that is it will skip over blank lines and comments, so we will still have multiple warnings.

**Comment by mysticatea** [comment]
@nzakas It checks whether the previous token of a statement node is included in the current unreachable range, in order to distinguish that a node is consecutive from the current range. So blank lines and comments are not a barrier. e.g.

```js
function foo() {
    return;

    a();  // ← ERROR: Unreachable code. (no-unreachable)

    b()   // ↑ ';' token is included in the unreachable code, so this statement will be merged.
    // comment
    c();  // ↑ ')' token is included in the unreachable code, so this statement will be merged.
}
```

This is a screenshot of a prototype.

![image](https://cloud.githubusercontent.com/assets/1937871/16705525/614f1dec-45c7-11e6-876d-ff35c17de371.png)

![image](https://cloud.githubusercontent.com/assets/1937871/16705552/1b4158f0-45c8-11e6-9617-4ba312a3e639.png)

----

If there are unreachable codes non-consecutive, there are multiple warnings:

![image](https://cloud.githubusercontent.com/assets/1937871/16705544/edbbbbb4-45c7-11e6-8000-c625e627eb3d.png)

**Comment by adambuczynski** [comment]
That looks good to me

**Comment by nzakas** [comment]
@mysticatea that looks good. Is that with a new option or by default?

**Comment by mysticatea** [comment]
I think it's good for default behavior.
CLI does not show the range of problems, but people can distinguish whether unreachable code(s) exist or not.

**Comment by nzakas** [comment]
Sounds good. We should make sure to update the docs to make it clear what this new behavior is. 

@eslint/eslint-team we need a couple more 👍s here.

**Comment by markelog** [comment]
SGTM

**Comment by ilyavolodin** [comment]
:+1:

**Comment by nzakas** [comment]
Assuming that @markelog's "SGTM" is the same as :+1:, marking as accepted.

**Comment by adambuczynski** [comment]
Thanks guys :tada:
